### PR TITLE
feat(pod): add configmapdelete experiment action for pod ConfigMap de…

### PR DESCRIPTION
**Pull request by pupil-xwb**
…letion

- Add ConfigMapDeleteActionSpec and ConfigMapDeleteActionExecutor that deletes a ConfigMap depended on by a Pod, then restarts the Pod to simulate startup failure caused by missing ConfigMap
- Backup the original ConfigMap as a new resource with chaosblade.io labels and annotations before deletion; restore it on experiment destroy
- Support --configmap-name flag to target a specific ConfigMap, or auto-select the first non-optional ConfigMap from the Pod spec
- Register configmapdelete action in pod self-exp model with long description and CLI examples
- Add unit tests covering create/destroy flows, backup/restore, edge cases (already deleted, missing backup, etc.)
- Add example YAML files for configmap-delete experiment scenarios

Change-Id: Ie9a8b7430e6e99ecdfd6464b1adfeb10c9be85b1
Co-developed-by: Qoder <noreply@qoder.com>

### Describe what this PR does / why we need it
新增 configmapdelete 实验动作，支持删除 Pod 依赖的 ConfigMap 并重启 Pod，模拟因 ConfigMap 缺失导致的 Pod 启动失败场景。实验销毁时自动恢复原始 ConfigMap。

### Does this pull request fix one issue?
NONE

### Describe how you did it
- 新增 ConfigMapDeleteActionSpec 和 ConfigMapDeleteActionExecutor，实现 create（备份 ConfigMap → 删除 ConfigMap → 删除 Pod 触发重启失败）和 destroy（从备份恢复原始 ConfigMap）流程
- 支持 --configmap-name 参数指定目标 ConfigMap；未指定时自动选择 Pod spec 中第一个非 optional 的 ConfigMap
- 备份 ConfigMap 以带有 chaosblade.io 标签和注解的新资源形式保存，包含原始名称、命名空间、标签和注解信息
- 在 pod.go 中注册 configmapdelete action，并添加 LongDesc 和 CLI 示例
- 新增单元测试，覆盖 create/destroy 正常流程、备份恢复、ConfigMap 已删除、备份缺失等边界场景
- 新增 3 个示例 YAML 文件

### Describe how to verify it
0. 运行单元测试：go test ./exec/pod/ -run TestConfigMapDelete -v
1. 在 K8s 集群中部署一个依赖 ConfigMap 的 Pod
2. 执行 blade create k8s pod-pod configmapdelete --labels "app=test" --namespace default --kubeconfig ~/.kube/config
3. 验证目标 ConfigMap 被删除、Pod 被重启且处于启动失败状态
4. 执行 destroy 命令，验证 ConfigMap 被恢复、Pod 恢复正常

### Special notes for reviews
- 备份 ConfigMap 的名称通过 SHA256 哈希生成，避免超出 K8s 253 字符名称限制
- destroy 阶段恢复 ConfigMap 时会还原原始的 labels 和 annotations并重启 Pod

### CODE REVIEW

File: exec/pod/configmapdeleteexp.go [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+				}
+				status = status.CreateFailResourceStatus(
+					fmt.Sprintf("delete configmap %s failed: %v", resolvedCMName, err), spec.K8sExecFailed.Code)
+				statuses = append(statuses, status)
+				continue
+			}
+			logrusField.Infof("deleted configmap %s/%s", c.Namespace, resolvedCMName)
+
+			// Mark as processed only after both backup and delete succeed
+			processedCMs[cmKey] = true
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go]: 已修复。将去重检查 processedCMs[cmKey] 提前到 Get ConfigMap 之前，已处理的 CM 直接跳到删除 Pod 步骤，避免后续 Pod 因 NotFound 被误标为失败。

File: exec/pod/configmapdeleteexp.go [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+	var experimentStatus v1alpha1.ExperimentStatus
+	if success {
+		experimentStatus = v1alpha1.CreateSuccessExperimentStatus(statuses)
+	} else {
+		experimentStatus = v1alpha1.CreateFailExperimentStatus("see resStatuses for details", statuses)
+	}
+	return spec.ReturnResultIgnoreCode(experimentStatus)
+}
+
+// destroy restores the backed-up ConfigMap and deletes the Pod to trigger a healthy restart.
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:240]: 感谢指出。代码行为是正确的（destroy 时需要删除 Pod 触发正常重建），是 PR 描述写错了。已更新 PR 描述的 Special notes 部分。

File: exec/pod/configmapdeleteexp.go [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+				continue
+			}
+
+			if err := d.deleteBackupConfigMap(ctx, backupCM); err != nil {
+				logrusField.Warningf("delete backup configmap %s/%s failed: %v", ns, backupCM.Name, err)
+			}
+		}
+	}
+
+	// Delete all matched Pods to trigger healthy rebuilds
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:303]: 感谢指出。代码行为是正确的（destroy 时需要删除 Pod 触发正常重建），是 PR 描述写错了。已更新 PR 描述的 Special notes 部分。

File: exec/pod/configmapdeleteexp.go [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+			Identifier: c.GetIdentifier(),
+		}
+
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      c.PodName,
+				Namespace: c.Namespace,
+			},
+		}
+		if err := d.client.Delete(ctx, pod); err != nil {
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:318]: 感谢指出。代码行为是正确的（destroy 时需要删除 Pod 触发正常重建），是 PR 描述写错了。已更新 PR 描述的 Special notes 部分。

File: exec/pod/configmapdeleteexp.go [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+
+	if allSuccess {
+		return spec.ReturnResultIgnoreCode(v1alpha1.CreateDestroyedExperimentStatus(statuses))
+	}
+	return spec.ReturnResultIgnoreCode(v1alpha1.CreateFailExperimentStatus("see resStatuses for details", statuses))
+}
+
+// createBackupConfigMap creates a backup copy of the original ConfigMap.
+func (d *ConfigMapDeleteActionExecutor) createBackupConfigMap(ctx context.Context, experimentId string, originalCM *v1.ConfigMap) error {
+	backupName := getBackupConfigMapName(experimentId, originalCM.Namespace, originalCM.Name)
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:346]: 已修复。将完整 experimentId 加入 SHA256 哈希输入，现在备份名称为 hash(experimentId/namespace/cmName)，不同实验不会再产生相同的备份名称。

File: exec/pod/configmapdeleteexp.go [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+		BinaryData: copyByteMap(originalCM.BinaryData),
+	}
+
+	if err := d.client.Create(ctx, backupCM); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			logrus.WithField("experiment", experimentId).Infof("backup configmap %s already exists, skip creation", backupName)
+			return nil
+		}
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:390]: 已修复。将完整的experimentId加入SHA256存储输入，现在备份名称为hash(experimentId/namespace/cmName)，不同的实验不会再产生相同的名称备份。

File: exec/pod/configmapdeleteexp.go [hunk 7]
```diff
@@ -0,0 +1,576 @@
...
+
+	// Restore original labels
+	var originalLabels map[string]string
+	if labelsJSON := backupCM.Annotations[ChaosBladeOriginalLabelsAnn]; labelsJSON != "" {
+		_ = json.Unmarshal([]byte(labelsJSON), &originalLabels)
+	}
+	// Restore original annotations
+	var originalAnnotations map[string]string
+	if annsJSON := backupCM.Annotations[ChaosBladeOriginalAnnsAnn]; annsJSON != "" {
+		_ = json.Unmarshal([]byte(annsJSON), &originalAnnotations)
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go]: 已修复。json.Unmarshal 失败时现在会记录 logrus.Warning 日志，同时备份阶段的 json.Marshal 也加了同样的日志处理。恢复流程不中断，因为恢复 Data 比元数据更重要，且备份 JSON 由我们自身 Marshal 生成，损坏概率极低。

File: exec/pod/configmapdeleteexp.go [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+}
+
+// ConfigMapRef describes a ConfigMap reference found in a Pod spec.
+type ConfigMapRef struct {
+	Name     string
+	Optional bool
+	Source   string // "volume", "envFrom", "envValueFrom"
+}
+
+// collectConfigMapReferences scans a Pod spec and returns all ConfigMap references.
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:468]: 感谢指出。代码实现是符合预期的（应当扫描 Volume 挂载的 ConfigMap 且优先选择），是 PR 描述遗漏了。已更新 PR 描述。

File: exec/pod/configmapdeleteexp.go [hunk 9]
```diff
@@ -0,0 +1,576 @@
...
+		seen[name] = len(refs)
+		refs = append(refs, ConfigMapRef{Name: name, Optional: optional, Source: source})
+	}
+
+	// Scan volumes first (highest priority for auto-selection)
+	for _, vol := range pod.Spec.Volumes {
+		if vol.ConfigMap != nil {
+			addRef(vol.ConfigMap.LocalObjectReference.Name, isOptional(vol.ConfigMap.Optional), "volume")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:492]: 感谢指出。代码实现是符合预期的（应当扫描 Volume 挂载的 ConfigMap 且优先选择），是 PR 描述遗漏了。已更新 PR 描述。

File: exec/pod/configmapdeleteexp.go [hunk 10]
```diff
@@ -0,0 +1,576 @@
...
+
+	// Scan all containers (regular + init)
+	allContainers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go]: 已修复。删除了自定义的 contains/searchString 函数，直接使用已导入的 strings.Contains。

File: exec/pod/configmapdeleteexp.go [hunk 11]
```diff
@@ -0,0 +1,576 @@
...
+	hash := sha256.Sum256([]byte(fmt.Sprintf("%s/%s", namespace, cmName)))
+	hashStr := fmt.Sprintf("%x", hash[:4])
+
+	expIdPrefix := experimentId
+	if len(expIdPrefix) > 8 {
+		expIdPrefix = expIdPrefix[:8]
+	}
+
+	return fmt.Sprintf("chaosblade-backup-%s-%s", expIdPrefix, hashStr)
+}
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:562]: 已修复。将完整 experimentId 加入 SHA256 哈希输入，现在备份名称为 hash(experimentId/namespace/cmName)，不同实验不会再产生相同的备份名称。

File: exec/pod/configmapdeleteexp.go [hunk 12]
```diff
@@ -0,0 +1,576 @@
...
+
+func (d *ConfigMapDeleteActionExecutor) Exec(uid string, ctx context.Context, expModel *spec.ExpModel) *spec.Response {
+	if _, ok := spec.IsDestroy(ctx); ok {
+		return d.destroy(ctx, expModel)
+	}
+	return d.create(ctx, expModel)
+}
+
+// create backs up the target ConfigMap, deletes it, then deletes the Pod to trigger a restart failure.
+func (d *ConfigMapDeleteActionExecutor) create(ctx context.Context, expModel *spec.ExpModel) *spec.Response {
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:112]: 现有测试已覆盖了核心辅助逻辑，不阻塞PR流程

File: exec/pod/configmapdeleteexp.go [hunk 13]
```diff
@@ -0,0 +1,576 @@
...
+	if success {
+		experimentStatus = v1alpha1.CreateSuccessExperimentStatus(statuses)
+	} else {
+		experimentStatus = v1alpha1.CreateFailExperimentStatus("see resStatuses for details", statuses)
+	}
+	return spec.ReturnResultIgnoreCode(experimentStatus)
+}
+
+// destroy restores the backed-up ConfigMap and deletes the Pod to trigger a healthy restart.
+func (d *ConfigMapDeleteActionExecutor) destroy(ctx context.Context, expModel *spec.ExpModel) *spec.Response {
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp.go:241]: 现有测试已覆盖了核心辅助逻辑，不阻塞PR流程

File: exec/pod/configmapdeleteexp_test.go [hunk 14]
```diff
@@ -0,0 +1,659 @@
...
+
+func searchString(s, substr string) bool {
+	for i := 0; i <= len(s)-len(substr); i++ {
+		if s[i:i+len(substr)] == substr {
+			return true
+		}
+	}
+	return false
+}
+
```
[the remark this answers is not in the document]
Comment (pupil-xwb) [reply exec/pod/configmapdeleteexp_test.go]: 已修复。删除了自定义的 contains/searchString 函数，直接使用已导入的 strings.Contains。

**Review by xcaspar — APPROVED** [review]
